From 4520726d7bc897c798743206821982f8bb98ed22 Mon Sep 17 00:00:00 2001
From: Chen Caidy <chen@caidy.cc>
Date: Wed, 19 Jul 2023 11:51:42 +0000
Subject: [PATCH] net: gptp: fix follow_up message timestamp

According to IEEE802.1AS 11.4.4.2.1, we need fill
preciseOriginTimestamp as syncEventEgressTimestamp.

In this follow_up message, prec_orig_ts need to filled
from net_pkt_timestamp(sync) for best accuracy. state machine
is software trigger with insufficient precision.

After this change, a grand master endpoint sync accuracy
increase from 3.5ms to 580ns with mimxrt1050_evk board.

Signed-off-by: Chen Caidy <chen@caidy.cc>
Upstream-Status: Backport [main: 762df873b4bc10a2a77e02dd2a58f292e78a191b]
---
 subsys/net/l2/ethernet/gptp/gptp_md.c       |  7 -------
 subsys/net/l2/ethernet/gptp/gptp_messages.c | 12 ++++++++++++
 2 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/subsys/net/l2/ethernet/gptp/gptp_md.c b/subsys/net/l2/ethernet/gptp/gptp_md.c
index 750998ca40..5c53df4fb3 100644
--- a/subsys/net/l2/ethernet/gptp/gptp_md.c
+++ b/subsys/net/l2/ethernet/gptp/gptp_md.c
@@ -61,13 +61,6 @@ static void gptp_md_follow_up_prepare(struct net_pkt *pkt,
 
 	hdr->log_msg_interval = sync_send->log_msg_interval;
 
-	fup->prec_orig_ts_secs_high =
-		htons(sync_send->precise_orig_ts._sec.high);
-	fup->prec_orig_ts_secs_low =
-		htonl(sync_send->precise_orig_ts._sec.low);
-	fup->prec_orig_ts_nsecs =
-		htonl(sync_send->precise_orig_ts.nanosecond);
-
 	fup->tlv_hdr.type = htons(GPTP_TLV_ORGANIZATION_EXT);
 	fup->tlv_hdr.len = htons(sizeof(struct gptp_follow_up_tlv));
 	fup->tlv.org_id[0] = GPTP_FUP_TLV_ORG_ID_BYTE_0;
diff --git a/subsys/net/l2/ethernet/gptp/gptp_messages.c b/subsys/net/l2/ethernet/gptp/gptp_messages.c
index 615df286ea..1aa6b64a38 100644
--- a/subsys/net/l2/ethernet/gptp/gptp_messages.c
+++ b/subsys/net/l2/ethernet/gptp/gptp_messages.c
@@ -238,8 +238,10 @@ struct net_pkt *gptp_prepare_sync(int port)
 struct net_pkt *gptp_prepare_follow_up(int port, struct net_pkt *sync)
 {
 	struct gptp_hdr *hdr, *sync_hdr;
+	struct gptp_follow_up *fup;
 	struct net_if *iface;
 	struct net_pkt *pkt;
+	struct net_ptp_time *sync_ts;
 
 	NET_ASSERT(sync);
 	NET_ASSERT((port >= GPTP_PORT_START) && (port <= GPTP_PORT_END));
@@ -255,7 +257,9 @@ struct net_pkt *gptp_prepare_follow_up(int port, struct net_pkt *sync)
 	net_pkt_set_priority(pkt, NET_PRIORITY_IC);
 
 	hdr = GPTP_HDR(pkt);
+	fup = GPTP_FOLLOW_UP(pkt);
 	sync_hdr = GPTP_HDR(sync);
+	sync_ts = net_pkt_timestamp(sync);
 
 	/*
 	 * Header configuration.
@@ -280,6 +284,14 @@ struct net_pkt *gptp_prepare_follow_up(int port, struct net_pkt *sync)
 	hdr->reserved1 = 0U;
 	hdr->reserved2 = 0U;
 
+	/*
+	 * Get preciseOriginTimestamp from previous sync message
+	 * according to IEEE802.1AS 11.4.4.2.1 syncEventEgressTimestamp
+	 */
+	fup->prec_orig_ts_secs_high = htons(sync_ts->_sec.high);
+	fup->prec_orig_ts_secs_low = htonl(sync_ts->_sec.low);
+	fup->prec_orig_ts_nsecs = htonl(sync_ts->nanosecond);
+
 	/* PTP configuration will be set by the MDSyncSend state machine. */
 
 	return pkt;
-- 
2.25.1

